File: 02. Algorithms/04. Recursion/Problems/02. 344. Reverse String/Solution.java
// https://leetcode.com/problems/reverse-string/
class Solution {
    public void reverseString(char[] s) {
        reverse(s, 0, s.length - 1);
    }
    
    void reverse(char[] s, int i, int j) {
        if(i >= j)
            return;
        char temp = s[i];
        s[i] = s[j];
        s[j] = temp;
        reverse(s, i+1, j-1);
    }
}

File: 02. Algorithms/04. Recursion/Problems/02. 344. Reverse String/solution.js
/**
 * https://leetcode.com/problems/reverse-string/
 * Time O(n) | Space O(n)
 * @param {character[]} s
 * @return {void} Do not return anything, modify s in-place instead.
 */
var reverseString = function(s) {

    const dfs = (left, right) => {
        if(left >=  right) return;
        swap(s, left, right);
        dfs(left+1, right-1);
    }

    dfs(0, s.length - 1);
};

 function swap(s,left,right) {
     const temp = s[left];
     s[left] = s[right];
     s[right] = temp;
 }
